addons/website/models/website.py

```diff
@@ -241,11 +245,138 @@ def unlink(self):
         attachments_to_unlink.unlink()
         return super(Website, self).unlink()
 
-    def create_and_redirect_to_theme(self):
+    def create_and_redirect_configurator(self):
         self._force()
-        action = self.env.ref('website.theme_install_kanban_action')
+        action = self.env.ref('website.start_configurator_act_url')
         return action.read()[0]
 
+    # ----------------------------------------------------------
+    # Configurator
+    # ----------------------------------------------------------
+    def _iap_website_service(self, route, params):
+        website_service_endpoint = self.env['ir.config_parameter'].sudo().get_param('website.website_service_endpoint', DEFAULT_ENDPOINT)
+        endpoint = website_service_endpoint + route
+        return iap_tools.iap_jsonrpc(endpoint, params=params)
+
+    @api.model
+    def configurator_init(self):
+        r = {}
+        company_id = self.get_current_website().company_id
+        r['features'] = self.env['website.configurator.feature'].search_read([], ['name', 'description', 'type', 'icon', 'website_types_preselection'])
+        r['logo'] = False
+        if company_id.logo and company_id.logo != company_id._get_logo():
+            r['logo'] = company_id.logo.decode('utf-8')
+        r['industries'] = self._iap_website_service('/website/industries', {'lang': self.env.user.lang})['industries']
+        return r
+
+    @api.model
+    def configurator_recommended_themes(self, industry_name, palette):
+        domain = [('name', '=like', 'theme%'), ('name', 'not in', ['theme_default', 'theme_common'])]
+        client_themes = [theme['name'] for theme in request.env['ir.module.module'].search_read(domain, ['name'])]
+        params = {
+            'industry_name': industry_name,
+            'palette': palette,
+            'client_themes': client_themes,
+        }
+        return self._iap_website_service('/website/recommended_themes', params)
+
+    @api.model
+    def configurator_skip(self):
+        self.configurator_done = True
+
+    @api.model
+    def configurator_apply(self, data):
+        def set_colors(selected_palette):
+            url = '/website/static/src/scss/options/user_values.scss'
+            custo = {'color-palettes-name': "'%s'" % selected_palette}
```

Comment:
- is this always safe? no verification needed for what you pass to `make_scss_customization()`?
- I think that `make_scss_customization` should be made private - is it ever called from the JS side?

Reply:
Which safety issue do you see? Code injection?
selected_palette is always a valid palette since the JS retrieved them directly from the style.

Reply:
Never thought about security issues with scss customization, I don't see what could happen besides breaking the CSS generation.

> I think that make_scss_customization should be made private - is it ever called from the JS side?

There is most probably something to do on that side though but maybe outside of this PR. `make_scss_customization` is indeed not called from the JS side... but there is a route dedicated to call it (so we could call the method directly and get rid of the route).. The scss customization was only done by routes directly before, then the utilities were moved to an abstract model... but for some reason we kept the routes, calling the abstract model method. I don't know if there was a reason or if it was just a big mistake. In any case, if something is to be done, it is getting rid of the extra routes and keeping these method public and call them directly if I am not mistaken.

Reply:
> I don't see what could happen besides breaking the CSS generation.

Isn't that enough? In this case, it looks ok because of the ACL on website and asset. Anyway making the methods private is always a good idea to reduce the risk of having a public method too trustful.

Reply:
> Isn't that enough?

If it was possible to do as a public user obviously yes... but like you said, there are access rights to do what these methods do (creating/updating views (well, now ir.asset records) and attachments). I may be wrong but I would trust what admins send to their own server... It seems to be the same as "we don't sanitize the HTML code which comes from the editor editions".

> Anyway making the methods private is always a good idea to reduce the risk of having a public method too trustful.

Like I explained, those are still methods called by dedicated routes. If we remove the extra routes which are probably there because we did not think about calling the abstract model methods directly via RPC, we'll want those methods to be public and rely on access rights. Tell me if am wrong, if not I'll make the PR to do that asap.

Reply:
Note I am all for extra checks to what we sent to methods even without security risk in mind: but what kind of check would you do here? "is the given color palette name a string?" We can't know more about the validity of the received value. If we want better and more robust, we'll have to consider this enormous task which I'd be happy to do: https://www.odoo.com/web#id=2463604&action=4043&model=project.task&view_type=form&cids=1&menu_id=4720